hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/discovery/HazelcastCloudDiscovery.java

```diff
@@ -0,0 +1,135 @@
+        HttpURLConnection httpConnection = (HttpURLConnection) url.openConnection();
+        httpConnection.setRequestMethod("GET");
+        httpConnection.setConnectTimeout(DEFAULT_TIMEOUT_MILLIS);
+        httpConnection.setReadTimeout(DEFAULT_TIMEOUT_MILLIS);
+        httpConnection.setRequestProperty("Accept-Charset", "UTF-8");
+        httpConnection.connect();
+        checkError(httpConnection);
+        return parseResponse(httpConnection.getInputStream());
+    }
+
+    private Map<Address, Address> parseResponse(InputStream is) throws IOException, URISyntaxException {
+
+        JsonValue jsonValue = Json.parse(readInputStream(is));
+        List<JsonValue> response = jsonValue.asArray().values();
+
+        Map<Address, Address> privateToPublicAddresses = new HashMap<Address, Address>();
+        for (JsonValue value : response) {
+            String privateAddress = value.asObject().get(PRIVATE_ADDRESS_PROPERTY).asString();
+            String publicAdress = value.asObject().get(PUBLIC_ADDRESS_PROPERTY).asString();
+            publicAdress = StringUtil.isNullOrEmptyAfterTrim(publicAdress) ? privateAddress : publicAdress;
+
+            Address publicAddr = createAddress(publicAdress);
+            privateToPublicAddresses.put(new Address(privateAddress, publicAddr.getPort()), publicAddr);
+        }
+
+        return privateToPublicAddresses;
+    }
+
+    private Address createAddress(String hostname) throws IOException, URISyntaxException {
+        URI uri = new URI("dummy://" + hostname);
+        String host = uri.getHost();
+        int port = uri.getPort();
+        return new Address(host, port);
+    }
+
+    private static String readInputStream(InputStream is) throws IOException {
+        BufferedReader in = new BufferedReader(new InputStreamReader(is, "UTF-8"));
+        String line;
+        StringBuilder response = new StringBuilder();
+        while ((line = in.readLine()) != null) {
+            response.append(line);
+        }
+        in.close();
+        return response.toString();
+    }
+
+    private static void checkError(HttpURLConnection httpConnection) throws IOException {
+        int responseCode = httpConnection.getResponseCode();
+        if (isError(responseCode)) {
+            String errorMessage = extractErrorMessage(httpConnection);
+            throw new IOException(errorMessage);
+        }
+    }
+
+    private static String extractErrorMessage(HttpURLConnection httpConnection) {
+        InputStream errorStream = httpConnection.getErrorStream();
+        return errorStream == null ? "" : readFrom(errorStream);
+    }
+
+    private static String readFrom(InputStream stream) {
+        Scanner scanner = (new Scanner(stream, "UTF-8")).useDelimiter("\\A");
+        return scanner.hasNext() ? scanner.next() : "";
+    }
+
+    private static boolean isError(int responseCode) {
+        return responseCode >= MIN_HTTP_CODE_FOR_ERROR && responseCode < MAX_HTTP_CODE_FOR_ERROR;
```

Comment:
what about responseCode != 200 (OK)?

Reply:
This was a copy from aws implementation, it seems these codes are specific to aws
http://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html

I will replace it with suggested one.